src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,214 @@
+        private IDictionary<int, (IMappingTagSpan<InlineDiagnosticsTag> mapTagSpan, SnapshotSpan snapshotSpan)> GetSpansOnEachLine(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            if (changedSpanCollection.IsEmpty())
+            {
+                return SpecializedCollections.EmptyDictionary<int, (IMappingTagSpan<InlineDiagnosticsTag>, SnapshotSpan)>();
+            }
+
+            var map = new Dictionary<int, (IMappingTagSpan<InlineDiagnosticsTag> mapTagSpan, SnapshotSpan snapshotSpan)>();
+            var viewLines = TextView.TextViewLines;
+
+            foreach (var changedSpan in changedSpanCollection)
+            {
+                if (!viewLines.IntersectsBufferSpan(changedSpan))
+                {
+                    continue;
+                }
+
+                var tagSpans = TagAggregator.GetTags(changedSpan);
+                foreach (var tagMappingSpan in tagSpans)
+                {
+                    if (!ShouldDrawTag(changedSpan, tagMappingSpan))
+                    {
+                        continue;
+                    }
+
+                    var mappedPoint = GetMappedPoint(changedSpan, tagMappingSpan);
+
+                    // mappedPoint is known to not be null here because it is checked in the ShouldNotDrawTag method call.
+                    var lineNum = mappedPoint!.Value.GetContainingLine().LineNumber;
+
+                    // If the line does not have an associated tagMappingSpan and changedSpan, then add the first one.
+                    if (!map.TryGetValue(lineNum, out var value))
+                    {
+                        map.Add(lineNum, (tagMappingSpan, changedSpan));
+                    }
+
+                    // If the map has a value and the value is not a syntax error, then rewrite the value in the map since
+                    // each line with errors will be ordered by importance
+                    if (value.mapTagSpan is not null && value.mapTagSpan.Tag.ErrorType is not PredefinedErrorTypeNames.SyntaxError)
+                    {
+                        map[lineNum] = (tagMappingSpan, changedSpan);
+                    }
+                }
+            }
+
+            return map;
+        }
+
+        /// <summary>
+        /// Iterates through the mapping of line number to span and draws the diagnostic in the appropriate position on the screen,
+        /// as well as adding the tag to the adornment layer.
+        /// </summary>
+        protected override void AddAdornmentsToAdornmentLayer_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            var viewLines = TextView.TextViewLines;
+            var map = GetSpansOnEachLine(changedSpanCollection);
+            foreach (var (lineNum, (tagMappingSpan, snapshotSpan)) in map)
```

Comment:
afaict, you only use this snapshotSpan to get a Snapshot below.  So having it be a snapshotSpan isn't useful.  Also, if this snapshotSpan just points at the snapshot the changedSpanCollection points to, then there's no need to compute and store this for each entry in the map.  You can just grab it off the changedSpanCollection.

Reply:
you're right, it was unnecessary.,